### PR TITLE
Advanced Effect Demo - Second UV set

### PR DESCRIPTION
**Pull request by swingingtom**
this PR could start discussions about advanced effects with three-mesh-ui and provided demos. 
Not really sure this could really be useful as example itself rn.

edit: following #112

### CODE REVIEW

File: examples/second_set_uvs.js [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        textureC.needsUpdate = true;
+
+        const textContent = `Three Mesh-Ui`.toUpperCase();
+        const text = new ThreeMeshUI.Text({
+            content: textContent,
+        });
+        container.add(
+            text
+        );
+
```
Comment (swingingtom) [inline examples/second_set_uvs.js:108]: The demo/sample part starts here

File: examples/second_set_uvs.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+                // add a second set of UVs as geometry attributes
+                // Gemeotry Attributes are automatically passed to vertexShader
+                text.children[0].geometry.setAttribute('uvC', new BufferAttribute(new Float32Array(secondUVSetArray), 2));
+
+
+                // 4. Replace the current fontMaterial shader with a customized one
+                text.fontMaterial.uniforms.u_map = { value: textureC };
+                text.fontMaterial.uniforms.u_uvCTransform = {value:textureC.matrix};
+
```
Comment (swingingtom) [inline examples/second_set_uvs.js:161]: I didn't really thought a lot about how to reuse existing shaders. Instead they are currently copy pasted from three-mesh-ui MaterialManager with small additions

**Comment by swingingtom** [comment]
Parts : 
```javascript
// 1. BOX REFERENCE
// 2. BOX DIMENSIONS
// 3. SECOND UV SET
```
Might be automated from three-mesh-ui without any core changes on current components, but with something like a runtime mixin.

```javascript
const text = new ThreeMeshUI.Text({content:"Hello World"});
// then
ThreeMeshUI.extras.UVCMixin( text );
```
where `ThreeMeshUI.extras.UVCMixin` would be something like : 
```javascript
ThreeMeshUI.extras.UVCMixin = function( textComponent ){
   const originalCreateText = textComponent.createText;

   textComponent.createText = function(){
        const mesh =  originalCreateText();

        // 1. BOX REFERENCE
        mesh.geometry.computeBoundingBox();
        const bBox = mesh.geometry.boundingBox;

       // 2. BOX DIMENSIONS
      const geometryW = bBox.max.x - bBox.min.x;
      const geometryH = bBox.max.y - bBox.min.y;

      // 3. SECOND UV SET
     const secondUVSetArray = [];
     const textVerticesArray = mesh.geometry.getAttribute('position').array;
     for (let i = 0; i < textVerticesArray.length; i += 3) {
          const vertX = textVerticesArray[i];
          const vertY = textVerticesArray[i + 1];

          secondUVSetArray.push((vertX / geometryW) + 0.5);
          secondUVSetArray.push((vertY / geometryH) + 0.5);
      }
      mesh.geometry.setAttribute('uvC', new BufferAttribute(new Float32Array(secondUVSetArray), 2));

      return mesh;
   }
}
```
Even if those kind of structure could offer some ease from three-mesh-ui itself, I don't really like it much, because it's far from three way-of-doing-things. But at the end, users will save lines of codes for doing those, it will ease the learning curve of  doing advanced uses of three-mesh-ui which might be the real point.

**Comment by felixmariotto** [comment]
Mmh let's imagine that what you want to create the second set of uv for is the whole content of a `Block`, consisting of 3 `Texts` (of different colors) and one `InlineBlock` smiley, then wouldn't this structure be limited ?

I think the computation of the second set of uvs should be done at container-level, either with a functional API with code in a separate module as you proposed, like this:
```js
const container = new ThreeMeshUI.Block({ /* ... */ })
const text = new ThreeMeshUI.Text({ content: "Hello World" });
container.add( text );
ThreeMeshUI.extras.computeUV2( container );
```
Or with a declarative API and the code for computing the uv2 attribute would be in `InlineManager`, like this:
```js
const container = new ThreeMeshUI.Block({
  computeUV2: true
  /* ... */
})
const text = new ThreeMeshUI.Text({ content: "Hello World" });
container.add( text );
```

I quite prefer the second option because:
1. `InlineManager` [already knows](https://github.com/felixmariotto/three-mesh-ui/blob/4c52c892e78eb06bba6c69a930eb59109ab801bd/src/components/core/InlineManager.js#L159-L175) some of what you compute with a bounding box. (you can also sort the lines `width`)
2. Since `Block` knows when it updates and [when children are added to it](https://github.com/felixmariotto/three-mesh-ui/blob/4c52c892e78eb06bba6c69a930eb59109ab801bd/src/components/core/MeshUIComponent.js#L289-L300), and children can know their parent's `computeUV2` value, we could recompute the second set of uv automatically when it's needed. If you update a `Text` content or add/remove inlines, everything would update with no additional code. Harder for the user to mess up.

Either way we should give the option to use the container space as reference for normalized UV space. (bottom left of the `Block` is 0,0 instead of the bottom left of the content bounding box being 0,0)

**Comment by swingingtom** [comment]
> Mmh let's imagine that what you want to create the second set of uv for is the whole content of a Block, consisting of 3 Texts (of different colors) and one InlineBlock smiley, then wouldn't this structure be limited ?

That would indeed be a nice effect

As block has width/height we can already set its proportions to fit on textures.

> Either way we should give the option to use the container as reference for normalized UV space. (bottom left of the Block is 0,0 instead of the bottom left of the content bounding box being 0,0)

Let's start with that. in the future, if it needs even more flexibility, both ways, declarative and runtime mixin could upgrade their property value to use a targeted geometry, without breaking declaration.
```javascript
ThreeMeshUI.extras.computeUV2( container, null(self)|anotherBlock|anotherGeometry|anotherBox3 );
// or 
const container = new ThreeMeshUI.Block({
  computeUV2: true(self)|anotherBlock|anotherGeometry|anotherBox
})
```

How would you plan it ?

**Comment by felixmariotto** [comment]
> Let's start with that.

OK, does it fit your own need ?

> How would you plan it ?

I like the second option (declarative), I feel it's more user-friendly because the user does not have to worry about when to update the second set of uv. Also `InlineManager` seems to be a logical place to do this.

**Comment by swingingtom** [comment]
> OK, does it fit your own need ?

Except if I missed some documentation, but I was never able to create a Text that is not contained in a Block. So it will fit my current requirements.

Okay for the declarative way, Block starts to have a lot of properties on it. I started a gui, and some of those properties might be removed I guess ( src? ) and some acts a bit strange (borderOpacity) : 
```javascript
container = new ThreeMeshUI.Block({
    borderOpacity: 0
  });
console.log("borderOpacity", container.getBorderOpacity()); // undefined instead of 0.0
```

I will report those at the end of the gui journey.

And what about fontShader; stay with user-code for now ?

**Comment by felixmariotto** [comment]
> And what about fontShader; stay with user-code for now ?

You had a good point [here](https://github.com/felixmariotto/three-mesh-ui/discussions/112#discussioncomment-1910773), I think it's best to use `onBeforeCompile`.

**Comment by swingingtom** [comment]
Im struggling with this demo, because I try to manage ease and flexibility. And its not that easy.
While this might stay a bit on hold, I might propose a PR to ease the use of others Materials for fontMaterial property.

**Comment by felixmariotto** [comment]
> Im struggling with this demo, because I try to manage ease and flexibility. And its not that easy.
> While this might stay a bit on hold, I might propose a PR to ease the use of others Materials for fontMaterial property.

What are you struggling with exactly ?

**Comment by swingingtom** [comment]
Obsolete due to 7.x.x forecast

**Pull request closed without merging**